draft-irtf-cfrg-bls-signature.md

```diff
@@ -611,28 +609,32 @@ Definitions:
 - "BLS-SIG-KEYGEN-SALT-" is an ASCII string comprising 20 octets.
 
 Procedure:
-1. PRK = HKDF-Extract("BLS-SIG-KEYGEN-SALT-", IKM || I2OSP(0, 1))
-2. OKM = HKDF-Expand(PRK, key_info || I2OSP(L, 2), L)
-3. SK = OS2IP(OKM) mod r
-4. return SK
+1. salt = "BLS-SIG-KEYGEN-SALT-"
+2. SK = 0
+3. while SK == 0:
+4.     salt = H(salt)
+5.     PRK = HKDF-Extract(salt, IKM || I2OSP(0, 1))
+6.     OKM = HKDF-Expand(PRK, key_info || I2OSP(L, 2), L)
+7.     SK = OS2IP(OKM) mod r
+8. return SK
```

Comment:
Since you're taking such conservative approach, then why not recognize that some might have put the original subroutine in action and make it even more conservative by
```
while True:
    ...
    if SK != 0:
        return SK
    salt = H(salt)

````

Reply:
Just to make sure I understand: the suggestion here is to wait to compute H(salt) until after the first iteration, i.e., use the ASCII string itself as the salt in the first iteration?

The reason not to do this is that HKDF's security proof is written for the case of random salt with size equal to the output length of H. HKDF's guarantees *may* be slightly weaker when using a salt that is structured (ASCII text) and shorter (20 bytes vs. at least 32 for H's output), which is why we are RECOMMENDING this. But of course people with existing implementations are free to use them provided they don't return SK == 0, per the text immediately below where you quoted.